File: src/main/java/April2020/Day11_DiameterOfBinaryTree.java
package April2020;

public class Day11_DiameterOfBinaryTree {
    public static void main(String[] args) {
        TreeNode node = new TreeNode(1);
        node.left = new TreeNode(2);
        node.right = new TreeNode(3);
        node.left.left = new TreeNode(4);
        node.left.right = new TreeNode(5);

        System.out.println(diameterOfBinaryTree(node));
    }

    public static int diameterOfBinaryTree(TreeNode root) {
        if (root == null) return 0;

        int outer = getMaxLength(root.left) + getMaxLength(root.right);
        int inner = Math.max(diameterOfBinaryTree(root.left), diameterOfBinaryTree(root.right));
        return Math.max(outer, inner);
    }

    private static int getMaxLength(TreeNode node) {
        return (node == null) ? 0 : 1 + Math.max(getMaxLength(node.left), getMaxLength(node.right));
    }

    private static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int x) { val = x; }
    }
}

File: src/main/kotlin/ThirtyDayLeetCodingChallenge/Day11_DiameterOfBinaryTree.kt
import kotlin.math.max

fun main() {
    val node = TreeNode(1)
    node.left = TreeNode(2)
    node.right = TreeNode(3)
    node.left!!.left = TreeNode(4)
    node.left!!.right = TreeNode(5)

    println(diameterOfBinaryTree(node))
}

private fun diameterOfBinaryTree(root: TreeNode?): Int {
    if (root == null) return 0
    val outer = getMaxLength(root.left) + getMaxLength(root.right)
    val inner = max(diameterOfBinaryTree(root.left), diameterOfBinaryTree(root.right))
    return max(outer, inner)
}

private fun getMaxLength(node: TreeNode?): Int {
    return if (node == null) 0 else 1 + max(getMaxLength(node.left), getMaxLength(node.right))
}

internal class TreeNode internal constructor(var value: Int) {
    var left: TreeNode? = null
    var right: TreeNode? = null
}
